core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java

```diff
@@ -99,6 +103,14 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
     return emulateNullDirectionWithIsNull(node, nullsFirst, desc);
   }
 
+  @Override public boolean supportsImplicitTypeCoercion(RexCall node) {
+    if (!super.supportsImplicitTypeCoercion(node)) {
+      return false;
+    }
+    RexNode operand = node.getOperands().get(0);
+    return operand instanceof RexLiteral && !SqlTypeFamily.NUMERIC.contains(node.type);
+  }
+
```

Comment:
From this code, you mean we can strip the case only if:

- The first operand is a STRING literal
- The second operand is not a NUMERIC

This is weird, can you please conform the type coercion rules ? Especially for the operators we strip the cast from: `>` `>=` `<` `<=` `IS NOT DISTINCT FROM` `<>`

Reply:
From the stripCastFromString method we are passing only the operands that has explicit cast applied. So it can be any of the operands. Now, the above code means, we can strip the cast for BigQuery if : 

- The RexNode on which we have applied the cast is a STRING LITERAL.
- We are trying to cast it to anything apart from NUMERIC type. 

We are not comparing between datatypes of operands of the comparison operators.

I have checked the type coercion rules for the above conditions you have mentioned (except IS  NOT DISTINCT FROM, BiqQuery doesn't have this operator yet) . This is my observation, you can check this [sheet](https://docs.google.com/spreadsheets/d/1GJ_VuDY7GQS-aPbWf4EKj73dYCqaRaEqPTjmXLkPW_g/edit#gid=0) .

For example,
hire_date = '1996-01-01 00:00:00' is valid in BiqQuery. But,
hire_date = concat('1996-01-01 ','00:00:00') is not valid. Need casting.
(hire_date is a TIMESTAMP column here)

and 

For Numeric datatype, 
employee_id = '10' not valid
employee_id = concat('1','2') not valid
10 = '10' is not valid.

Reply:
Thanks for the explain, i think we should match the cases that we exactly know well, because the implicit type coercion rules are relevant with many sql contexts.